technical-documents/postmortems/2022-02-02-inflation-vuln.md

```diff
@@ -0,0 +1,281 @@
+files). The diff added 217 lines, and removed 149 lines.
+
+The PR was well scoped, and all of the changes were relevant according to its description:
+
+> Refactors the usage of OVM_ETH so we can get most remaining integration tests working again. Also
+> reworks `vm.UsingOVM` to be `rcfg.UsingOVM` where `rcfg` is a new package within the rollup
+> folder. Was required in order to avoid an import cycle.
+
+The PR was reviewed at least twice, with inline comments that indicate attention to detail, although
+the
+[comments in the buggy code](https://github.com/ethereum-optimism/optimism/pull/1363/files#diff-11f5b63c52e9c7c30e4e599f96f84db5f08121e8eb623aa1176c2801389487b9)
+itself were sparse, and fairly high level.
+
+Notably, the eventual fix to the bug was made in `instructions.go`, a **file which was completely
+untouched by the PR**.
+
+#### Code and Organizational context
+
+The PR #1363 was one small part of a major architectural update (which we refer as a 'regenesis') to
+Optimism. The regenesis removed the OVM contracts, and enabled EVM equivalence.
+
+The total size of the update can be seen in the
+[regenesis 0.5.0 PR](https://github.com/ethereum-optimism/optimism/pull/1594/commits), which
+included the commits from the PR above. This was a massive upgrade, as we can see from the size of
+the PR (36,311 lines added, 47,430 lines removed), which consumed the attention of our entire
+engineering team with a sense of urgency for several months.
+
+An additional factor contributing to this bug was the significant complexity of the
+[L2Geth](https://github.com/ethereum-optimism/optimism/tree/master/l2geth) codebase, which is a fork
+of [Geth](https://github.com/ethereum/go-ethereum). Geth itself is already a very complex codebase.
+The changes introduced to L2Geth in order to support the OVM made it much more complex, such that
+very few people properly understood how it worked.
+
+The changes made for this regenesis mostly removed this complexity, and moved the behavior of L2Geth
+closer that of Geth. Unfortunately L2Geth had already diverged significantly, and the abstractions
+of the OVM leaked in enough that a change made in one part of the code could have major consequences
+elsewhere.
+
+More specifically: the OVM used `OVM_ETH`, and ERC20 token rather than native ETH, meaning that
+account balances were no longer kept in the state trie. However the EVM's `SELFDESTRUCT` works by
+deleting the balance in the trie. In addition, `SELFDESTRUCT` was not implemented in the OVM, meaning
+it was not present to remind us that it needed updating in the EVM.
+
+#### Lack of auditing
+
+We did not have an audit on the changes made to the regenesis. The rationale for this was that:
+
+1. the changes were mostly deleting code and simplifying the system by removing the OVM, and
+2. the availability of qualified auditors was extremely constrained.
+
+#### Conclusion regarding the introduction of the bug
```

Comment:
An additional point worth noting here is that the issue could've been prevented with a review of expected EVM behavior for any opcodes impacted by the change. For example, the logic should've been:

1. This impacts the way that balances and value transfers happen in the EVM.
2. Several opcodes refer to balance and value transfer.
3. SELFDESTRUCT involves value transfer.
4. Does SELFDESTRUCT behave the same way after the change?

Reply:
WDYT an improved PR process would look like to ensure that happens? Would it require the author to explicitly list the other parts of the system that could be affected and why they believe its not an issue?

Reply:
Tried to incorporate that here: https://github.com/ethereum-optimism/optimism/pull/2246/files#diff-b740618b9dad4e1d02d8a2eb24e62f3a67cdfa72a8a113a137b9dfb4c6ed597fR158-R167